sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -103,8 +116,12 @@ case class TransformWithStateExec(
     val keyObj = getKeyObj(keyRow)  // convert key to objects
     ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
     val valueObjIter = valueRowIter.map(getValueObj.apply)
-    val mappedIterator = statefulProcessor.handleInputRows(keyObj, valueObjIter,
-      new TimerValuesImpl(batchTimestampMs, eventTimeWatermarkForLateEvents)).map { obj =>
+    val mappedIterator = statefulProcessor.handleInputRows(
+      keyObj,
+      valueObjIter,
+      new TimerValuesImpl(batchTimestampMs, eventTimeWatermarkForLateEvents),
```

Comment:
I agree it's super confusing, but conceptually, two watermark values are really a one watermark value with two phases, like following:

1. the operator starts with watermark as eventTimeWatermarkForLateEvents (continuation of watermark in previous microbatch)
2. the operator processes all inputs
3. the operator advances the watermark as eventTimeWatermarkForEviction
4. the operator handles eviction based on watermark advancement
 
like row1, row2, row3, ..., rowN, WT (watermark marker) in the input queue, if it's easier to understand based on record-to-record based streaming engine.

That said, conceptually, here the watermark value is correct, but in onTimer, the watermark value should be eventTimeWatermarkForEviction. But if, we concern that users would be confusing (despite the fact that this is correct) and we want to provide the single value, arguably, eventTimeWatermarkForEviction seems to be more proper. We drop late events already, and users don't get the chance to process them anyway. So the focus should be more about eviction.

Reply:
Done - updated